### ISSUE TITLE
AWS::CloudWatch::Alarm - allow tags

### ISSUE THREAD

**Issue by Limess** [aws-cloudformation/cloudformation-coverage-roadmap#64]
## 1. Title 

CloudWatch supports tagging and untagging alarms, and it should be supported in CloudFormation.

## 2. Scope of request

Allow for tags on create, updating tags, etc.

## 3. Expected behavior

See 2 above.

## 4. Test case recommendation

* Adding tags on create of CloudWatch alarm.
* Addings tags on update of CloudWach alarm.
* Removing tags on update of CloudWatch Alarm.

## 5. Links to existing API doc

API Docs: https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_TagResource.html

> In CloudWatch, alarms can be tagged.

## 6. Category tag

Management 

## 7. Any additional context

The documentation on this feature is sparse.

Existing resource: [AWS::CloudWatch::Alarm](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-cw-alarm.html).

This is supported by Boto: [CloudWatch.Client.tag_resource](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/cloudwatch.html#CloudWatch.Client.tag_resource).

**Comment by tvb** [comment]
@luiseduardocolon It has been sitting in status "We're working on it" since Jan 11th. Any update for this? My finance dept. is killing me for not being able to filter on tags for our alarms.

**Comment by JanneKataja-TomTom** [comment]
Also hoping this gets implemented soon. We want to set incident priority and urgency based on CloudWatch alarm tags, and not have to configure these separately in another system.

**Comment by qmg-drettie** [comment]
@luiseduardocolon any updates on this ?

**Comment by thomasklinger1234** [comment]
> Also hoping this gets implemented soon. We want to set incident priority and urgency based on CloudWatch alarm tags, and not have to configure these separately in another system.

THIS. It would be so much easier to tag alarms and then create a CloudWatch event rule which forwards alarm state changes to a Lambda which checks the tags and forwards messages.. Thats my plan once tagging is available

**Comment by alexbarnes** [comment]
Argh. Just ran into this one.

**Comment by djfurman** [comment]
Adding my plus-one here. We have so many teams and applications we don't know who to contact about cleaning up their alarms that point to resources that may no longer exist.

**Comment by butangero** [comment]
+1 as well - this would be huge for us to be able to fully utilize tags for automating dashboard creation.

**Comment by lisakester** [comment]
+1 We need cloudwatch alarm tags so we can allow people to tag their alarms with information that's needed for our shared lambda functions.

**Comment by ptdel** [comment]
Hoping to implement the following in one go: https://aws.amazon.com/about-aws/whats-new/2019/04/you-can-now-use-resource-level-permissions-for-amazon-cloudwatch/ would be nice to deploy tagged cloudwatch alarms via CFN alongside the policy that applies to them.

**Comment by jenshoffmann1331** [comment]
using description for tags. so desperate we are.

**Comment by tkrajca** [comment]
Just ran into this one as well :( any ETAs?

**Comment by farski** [comment]
If anyone needs a solution to this before it's supported natively, this is the custom resource Lambda that I'm using:

```javascript
const response = require('cfn-response');
const AWS = require('aws-sdk');
const cloudwatch = new AWS.CloudWatch();

exports.handler = async (event, context) => {
  console.log(event);

  try {
    const tags = event.ResourceProperties.Tags;
    const arn = event.ResourceProperties.AlarmArn;

    if (!tags || !tags.length || !arn) {
      console.error('AlarmArn and Tags properties must be defined');
      await response.send(event, context, response.FAILED, {});
    }

    if (event.RequestType === 'Create') {
      // Create all tags on the custom resource
      await cloudwatch.tagResource({
        ResourceARN: arn,
        Tags: tags,
      }).promise();
    } else if (event.RequestType === 'Update') {
      // Remove tags that were present in the old resource properties, but are
      // no longer present
      const previousTags = event.OldResourceProperties.Tags;
      const currentTagKeys = tags.map(t => t.Key);

      const staleTags = previousTags.filter(p => !currentTagKeys.includes(p.Key));

      if (staleTags.length) {
          await cloudwatch.untagResource({
            ResourceARN: arn,
            TagKeys: staleTags.map(t => t.Key),
          }).promise();
      }

      // Create/update all values present in the current resource properties
      await cloudwatch.tagResource({
        ResourceARN: arn,
        Tags: tags,
      }).promise();
    } else if (event.RequestType === 'Delete') {
      // Remove all tags on the custom resource
      await cloudwatch.untagResource({
        ResourceARN: arn,
        TagKeys: tags.map(t => t.Key),
      }).promise();
    }

    await response.send(event, context, response.SUCCESS, {});
  } catch (error) {
    console.error(error);
    await response.send(event, context, response.FAILED, {});
  }
};
```

Which would work with a custom resource like this: 

```yaml
  MyAlarmTags:
    Type: Custom::CloudWatchAlarmTags
    Properties:
      ServiceToken: !Ref CloudWatchAlarmTaggerServiceToken
      AlarmArn: !GetAtt MyAlarm.Arn
      Tags:
        - { Key: Severity, Value: Critical }
        - { Key: LogGroup, Value: arn:aws:logs:us-east-1:1234567890:log-group:my-service-logs }
```

**Comment by venkiit58** [comment]
Is there any news about this? please.

**Comment by qguang** [comment]
+1

**Comment by ytsipun** [comment]
+1 from 2021!

**Comment by jrombi** [comment]
+1

**Comment by phene** [comment]
+1 from 2022

**Comment by pchaganti** [comment]
+1

**Comment by ryanzhu2017** [comment]
+1

**Comment by djfurman** [comment]
+1 from April 2022

**Comment by dougallaninthecloud** [comment]
+1

**Comment by gmahe** [comment]
+1

**Comment by JamesKyburz** [comment]
+1 from 2022

**Comment by towerbe** [comment]
+1 from August 2022.  Come on guys, this is basic stuff.  This issue has been open for 3+ years.

**Comment by teague-mfa** [comment]
+1

**Comment by jwalls89** [comment]
+1

**Comment by joshbfei** [comment]
+1

**Comment by Risae** [comment]
Can someone please take one for the team and apply at AWS to add this feature?

**Comment by towerbe** [comment]
@Risae , Oh, I have, trust me, under my Global Account persona.  Still doesn't seem to make a difference.

I will poke our account team on this one again and see if we can get any traction.

**Comment by danw-mpl** [comment]
If anyone is having an issue with the rate limiting and is using the `AwsCustomResource`, we can set the retry configuration via environment variables:

```typescript
import { Construct } from 'constructs';

import * as iam from 'aws-cdk-lib/aws-iam';
import * as cloudwatch from 'aws-cdk-lib/aws-cloudwatch';
import * as customResources from 'aws-cdk-lib/custom-resources';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { Duration } from 'aws-cdk-lib';

interface Tag {
    readonly Key: string;
    readonly Value: string;
}

export class AlarmTags extends Construct {
    /**
     * A construct that adds tags to CloudWatch alarms.
     *
     * Currently this functionality is not supported by CloudFormation
     * so we're using a custom resource to achieve this.
     *
     * @example
     *
     * alarms.forEach(alarm => {
     *     new AlarmTags(this, `AlarmTags-${alarm.node.id}`, alarm, [
     *         { key: 'abc', value: 'def' },
     *         { key: 'def', value: 'abc' },
     *     ])
     * });
     *
     * @see https://github.com/aws-cloudformation/cloudformation-coverage-roadmap/issues/64
     *
     * @param scope The CDK scope.
     * @param id The logical ID.
     * @param alarm The CloudWatch Alarm.
     * @param tags A list of Tag objects.
     */
    constructor(scope: Construct, id: string, alarm: cloudwatch.Alarm, tags: Tag[]) {
        super(scope, id);

        const sdkCall: customResources.AwsSdkCall = {
            service: 'CloudWatch',
            action: 'tagResource',
            physicalResourceId: customResources.PhysicalResourceId.of(`${alarm.alarmName}Tags`),
            parameters: {
                ResourceARN: alarm.alarmArn,
                Tags: tags,
            },
        };

        const cr = new customResources.AwsCustomResource(this, `id-${alarm.node.id}`, {
            resourceType: 'Custom::AlarmTags',
            onCreate: sdkCall,
            onUpdate: sdkCall,
            logRetention: logs.RetentionDays.ONE_DAY,
            policy: customResources.AwsCustomResourcePolicy.fromStatements([
                new iam.PolicyStatement({
                    actions: ['cloudwatch:TagResource'],
                    resources: [alarm.alarmArn],
                }),
            ]),
            timeout: Duration.minutes(5),
        });

        const fn = cr.node.findChild('Provider') as lambda.SingletonFunction;
        fn.addEnvironment('AWS_MAX_ATTEMPTS', '10');
        fn.addEnvironment('AWS_RETRY_MODE', 'adaptive');
    }
}
```

**Comment by Risae** [comment]
> @Risae , Oh, I have, trust me, under my Global Account persona. Still doesn't seem to make a difference.
> 
> I will poke our account team on this one again and see if we can get any traction.

I did the same....
But i just noticed that even Terraform supports tagging of CloudWatch Alarms: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/cloudwatch_composite_alarm

Anybody know why Terraform is better than CloudFormation?!

**Comment by Gekkko** [comment]
+1

**Comment by ngander-amfam** [comment]
+1 This is getting pretty ridiculous. Why no CloudFormation support after 3+ years?

**Comment by tonyshearer** [comment]
+1

**Comment by ellisium** [comment]
+1

**Comment by ellisium** [comment]
+1 from 2023

**Comment by Gtofig** [comment]
+1. Lack of tag support for alarms makes it hard for CDK users to apply tags consistently using CDK aspects

**Comment by rbartell** [comment]
+1

**Comment by wjgrainger** [comment]
+1

**Comment by zoltanmajo** [comment]
+1 from 2023. Thank you in advance.

**Comment by oros-fundmore** [comment]
+1

**Comment by felihol** [comment]
+1 on this, is there any plan to solve this somehow?

**Comment by larse514** [comment]
+1 running into this limitation with our CDK usage

**Comment by srinivas-vangari** [comment]
+1000000000

This is already supported in Terraform, why not in CFn. Other services do support this via CFn, why not CFn?

**Comment by Risae** [comment]
It is now possible to Tag `AWS::CloudWatch::Alarm` using

- The AWS CLI: https://awscli.amazonaws.com/v2/documentation/api/latest/reference/cloudwatch/tag-resource.html
- The AWS GUI: https://aws.amazon.com/about-aws/whats-new/2023/04/amazon-cloudwatch-console-capabilities-data-visualizations/

bout time for CloudFormation to support it too!

**Comment by ghettosamson** [comment]
I am a big user of CloudFormations and Tagging resources to track resources for prod/beta/demo and would really like this feature to be implemented for CloudWatch Alarms.

**Comment by corey-cole** [comment]
+1

**Comment by curtspiteri** [comment]
+1

**Comment by saketkumar-nasstar** [comment]
+1

**Comment by phancox** [comment]
+1

**Comment by alanzhaonys** [comment]
OMG is this available now?

**Comment by zpg-jlowe** [comment]
+1

**Comment by isuftin** [comment]
It's been 4 years....

...so thirsty...

**Comment by tbhangalenice** [comment]
We also have many alarms in CloudFormation that need to be tagged. Is there any update on this?

**Comment by zbarr** [comment]
+1

**Comment by ilons** [comment]
4 (!) years later, still no support for tagging of alarms using CloudFormation, _please_ add this so that we can start attributing cost :disappointed:

**Comment by FarrOut** [comment]
Just retested, confirmed this is still not implemented.

*CDK*
```
        metric = cloudwatch.Metric(
            namespace="MyNamespace",
            metric_name="MyMetric",
            dimensions_map={
                "ProcessingStep": "Download"
            }
        )

        alarm = cloudwatch.Alarm(self, "Alarm",
                                 metric=metric,
                                 threshold=100,
                                 evaluation_periods=2
                                 )
        Tags.of(alarm).add('ThisTag', 'Was added at resource-level') # <- This has no effect
```

The synthesized Cfn template does not add ```Tags``` property...

```
{
   "Type": "AWS::CloudWatch::Alarm",
   "Properties": {
    "ComparisonOperator": "GreaterThanOrEqualToThreshold",
    "Dimensions": [
     {
      "Name": "ProcessingStep",
      "Value": "Download"
     }
    ],
    "EvaluationPeriods": 2,
    "MetricName": "MyMetric",
    "Namespace": "MyNamespace",
    "Period": 300,
    "Statistic": "Average",
    "Threshold": 100
   }
```

Resulting resource obviously has no tags applied.

Manually adding ```Tags``` property...
```
        cfn_alarm = alarm.node.default_child
        cfn_alarm.add_property_override("Tags.0.Value", "Raw override")
```
...errors out with..
```
Properties validation failed for resource Alarm7103F465 with message: #: extraneous key [Tags] is not permitted
```

This is expected, as inspecting the resource type schema...
```
aws cloudformation describe-type --type-name AWS::CloudWatch::Alarm --type RESOURCE --output json > aws-cloudwatch-alarm.json
```
...reveals the following schema...
```
{
    "typeName": "AWS::CloudWatch::Alarm",
    "description": "Resource Type definition for AWS::CloudWatch::Alarm",
    "additionalProperties": false,
    "sourceUrl": "https://github.com/aws-cloudformation/aws-cloudformation-resource-providers-cloudwatch.git",
    "properties": {
      "ThresholdMetricId": {
        "description": "In an alarm based on an anomaly detection model, this is the ID of the ANOMALY_DETECTION_BAND function used as the threshold for the alarm.",
        "type": "string"
      },
      "EvaluateLowSampleCountPercentile": {
        "description": "Used only for alarms based on percentiles.",
        "type": "string"
      },
      "ExtendedStatistic": {
        "description": "The percentile statistic for the metric associated with the alarm. Specify a value between p0.0 and p100.",
        "type": "string"
      },
      "ComparisonOperator": {
        "description": "The arithmetic operation to use when comparing the specified statistic and threshold.",
        "type": "string"
      },
      "TreatMissingData": {
        "description": "Sets how this alarm is to handle missing data points. Valid values are breaching, notBreaching, ignore, and missing.",
        "type": "string"
      },
      "Dimensions": {
        "description": "The dimensions for the metric associated with the alarm. For an alarm based on a math expression, you can't specify Dimensions. Instead, you use Metrics.",
        "type": "array",
        "uniqueItems": false,
        "insertionOrder": false,
        "items": {
          "$ref": "#/definitions/Dimension"
        }
      },
      "Period": {
        "description": "The period in seconds, over which the statistic is applied.",
        "type": "integer"
      },
      "EvaluationPeriods": {
        "description": "The number of periods over which data is compared to the specified threshold.",
        "type": "integer"
      },
      "Unit": {
        "description": "The unit of the metric associated with the alarm.",
        "type": "string"
      },
      "Namespace": {
        "description": "The namespace of the metric associated with the alarm.",
        "type": "string"
      },
      "OKActions": {
        "description": "The actions to execute when this alarm transitions to the OK state from any other state.",
        "type": "array",
        "uniqueItems": false,
        "items": {
          "type": "string"
        }
      },
      "AlarmActions": {
        "description": "The list of actions to execute when this alarm transitions into an ALARM state from any other state.",
        "type": "array",
        "uniqueItems": false,
        "items": {
          "type": "string"
        }
      },
      "MetricName": {
        "description": "The name of the metric associated with the alarm.",
        "type": "string"
      },
      "ActionsEnabled": {
        "description": "Indicates whether actions should be executed during any changes to the alarm state. The default is TRUE.",
        "type": "boolean",
        "default": true
      },
      "Metrics": {
        "description": "An array that enables you to create an alarm based on the result of a metric math expression.",
        "type": "array",
        "uniqueItems": true,
        "items": {
          "$ref": "#/definitions/MetricDataQuery"
        }
      },
      "AlarmDescription": {
        "description": "The description of the alarm.",
        "type": "string"
      },
      "AlarmName": {
        "description": "The name of the alarm.",
        "type": "string"
      },
      "Statistic": {
        "description": "The statistic for the metric associated with the alarm, other than percentile.",
        "type": "string"
      },
      "InsufficientDataActions": {
        "description": "The actions to execute when this alarm transitions to the INSUFFICIENT_DATA state from any other state.",
        "type": "array",
        "uniqueItems": false,
        "items": {
          "type": "string"
        }
      },
      "Arn": {
        "description": "Amazon Resource Name is a unique name for each resource.",
        "type": "string"
      },
      "DatapointsToAlarm": {
        "description": "The number of datapoints that must be breaching to trigger the alarm.",
        "type": "integer"
      },
      "Threshold": {
        "description": "In an alarm based on an anomaly detection model, this is the ID of the ANOMALY_DETECTION_BAND function used as the threshold for the alarm.",
        "type": "number"
      }
    },
    "definitions": {
      "MetricStat": {
        "description": "This structure defines the metric to be returned, along with the statistics, period, and units.",
        "type": "object",
        "additionalProperties": false,
        "properties": {
          "Period": {
            "description": "The granularity, in seconds, of the returned data points.",
            "type": "integer"
          },
          "Metric": {
            "description": "The metric to return, including the metric name, namespace, and dimensions.",
            "$ref": "#/definitions/Metric"
          },
          "Stat": {
            "description": "The statistic to return.",
            "type": "string"
          },
          "Unit": {
            "description": "The unit to use for the returned data points.",
            "type": "string"
          }
        },
        "required": [
          "Stat",
          "Period",
          "Metric"
        ]
      },
      "Metric": {
        "description": "The Metric property type represents a specific metric.",
        "type": "object",
        "additionalProperties": false,
        "properties": {
          "MetricName": {
            "description": "The name of the metric.",
            "type": "string"
          },
          "Dimensions": {
            "description": "The dimensions for the metric.",
            "type": "array",
            "uniqueItems": false,
            "insertionOrder": false,
            "items": {
              "$ref": "#/definitions/Dimension"
            }
          },
          "Namespace": {
            "description": "The namespace of the metric.",
            "type": "string"
          }
        }
      },
      "Dimension": {
        "description": "Dimensions are arbitrary name/value pairs that can be associated with a CloudWatch metric.",
        "type": "object",
        "additionalProperties": false,
        "properties": {
          "Value": {
            "description": "The value for the dimension.",
            "type": "string"
          },
          "Name": {
            "description": "The name of the dimension.",
            "type": "string"
          }
        },
        "required": [
          "Value",
          "Name"
        ]
      },
      "MetricDataQuery": {
        "description": "This property type specifies the metric data to return, and whether this call is just retrieving a batch set of data for one metric, or is performing a math expression on metric data.",
        "type": "object",
        "additionalProperties": false,
        "properties": {
          "Label": {
            "description": "A human-readable label for this metric or expression.",
            "type": "string"
          },
          "MetricStat": {
            "description": "The metric to be returned, along with statistics, period, and units.",
            "$ref": "#/definitions/MetricStat"
          },
          "Id": {
            "description": "A short name used to tie this object to the results in the response.",
            "type": "string"
          },
          "ReturnData": {
            "description": "This option indicates whether to return the timestamps and raw data values of this metric.",
            "type": "boolean"
          },
          "Expression": {
            "description": "The math expression to be performed on the returned data.",
            "type": "string"
          },
          "Period": {
            "description": "The period in seconds, over which the statistic is applied.",
            "type": "integer"
          },
          "AccountId": {
            "description": "The ID of the account where the metrics are located, if this is a cross-account alarm.",
            "type": "string"
          }
        },
        "required": [
          "Id"
        ]
      }
    },
    "required": [
      "ComparisonOperator",
      "EvaluationPeriods"
    ],
    "handlers": {
      "create": {
        "permissions": [
          "cloudwatch:PutMetricAlarm",
          "cloudwatch:DescribeAlarms"
        ]
      },
      "update": {
        "permissions": [
          "cloudwatch:PutMetricAlarm",
          "cloudwatch:DescribeAlarms"
        ]
      },
      "delete": {
        "permissions": [
          "cloudwatch:DeleteAlarms",
          "cloudwatch:DescribeAlarms"
        ]
      },
      "list": {
        "permissions": [
          "cloudwatch:DescribeAlarms"
        ]
      },
      "read": {
        "permissions": [
          "cloudwatch:DescribeAlarms"
        ]
      }
    },
    "createOnlyProperties": [
      "/properties/AlarmName"
    ],
    "primaryIdentifier": [
      "/properties/AlarmName"
    ],
    "readOnlyProperties": [
      "/properties/Arn"
    ],
    "tagging": {
      "taggable": false
    }
  }
```

**Note**
- No ```Tags``` property
- ```taggable:false``` 

---

*tl;dr*
Still not supported.

**Comment by Risae** [comment]
`@cfn-github-issues-bot moved this from "We're working on it" to "Researching" in coverage-roadmap 3 days ago`

3 days ago was a sad day.

**Comment by brokenjpl** [comment]
Can we have this basic functionality yet?

**Comment by barnou** [comment]
Do you have any news on this issue ? It's really important to be able to tag resources we are creating using CloudFormation. It's possible to do it using Terraform, it should be also possible to do it by using the builtin product of AWS, shouldn't it ?

**Comment by Risae** [comment]
`@cfn-github-issues-bot moved this from "Researching" to "We're working on it" in coverage-roadmap 4 days ago`
Oh hey, they put it back to `We're working on it`...

**Comment by webmozart** [comment]
We also need this.

**Comment by theaussiepom** [comment]
And us too!

**Comment by LUKEALDRED** [comment]
Just ran across this. Are there any updates please? It's been 4 1/2 years! 🤦‍♂️

**Comment by LinmeiJ** [comment]
Any updates for adding tags to cloudwatch alarms via cloudformation?

**Comment by rdob2022** [comment]
Is there any update on this?

**Comment by eilgin** [comment]
I think the AWS team responsible for that (CFn? Cloudwatch?) will never implement this after all this time ...

**Comment by j-schwerer** [comment]
Hi,
thank you for the recent comments showing how this feature request matters to you. As you know, AWS strives to build our roadmap from customer feedback, and your comments here contribute as one of our sources of feedback.
With regards to this specific enhancement, we can assure you the CloudWatch team has not dismissed it from our roadmap and are willing to implement it.
We also apologize if the robot's status change caused any confusion, the topic is well defined in CloudWatch roadmap.
Please feel free to keep adding new information or comments here if needed.

**Comment by Risae** [comment]
> Hi, thank you for the recent comments showing how this feature request matters to you. As you know, AWS strives to build our roadmap from customer feedback, and your comments here contribute as one of our sources of feedback. With regards to this specific enhancement, we can assure you the CloudWatch team has not dismissed it from our roadmap and are willing to implement it. We also apologize if the robot's status change caused any confusion, the topic is well defined in CloudWatch roadmap. Please feel free to keep adding new information or comments here if needed.

Thank you, lets hope 2024 will be the year this issue can be closed!

**Comment by mwgamble** [comment]
I just wasted the best part of an hour digging through Cloudtrail logs trying to work out why I couldn't see any `cloudwatch:TagResoure` API calls coming from CloudFormation. Now I know why :(

**Comment by SubhanHakverdiyev** [comment]
We also need this!

**Comment by rmechi** [comment]
+1

**Comment by nsekulov** [comment]
+1

**Comment by peledies** [comment]
+1 come on AWS

**Comment by bhuvanprasadg** [comment]
+1 we needs this

**Comment by nickdavo** [comment]
I remember being at reInvent where AWS were talking about their differentiator being that they were customer obsessed.  Waiting 5 years for this does not feel customer obsessed.   Please add this soon.

**Comment by ninadvadujkar** [comment]
I just ran into this one as well :(

**Comment by bdoyle0182** [comment]
Pls

**Comment by cbertozzi** [comment]
+ 1

**Comment by guusvw** [comment]
Just for the record: +1

**Comment by ramakanth1997** [comment]
Any updates for adding tags to cloudwatch alarms via cloudformation?

**Comment by rgoltz** [comment]
Today, I run the following test via CloudFormation - Spoiler: I was able to add Tags via CFN to `AWS::CloudWatch::Alarm`.
Here my example in eu-central-1:

### Test Cases with Stack-Level-Tags
1) Create a new stack with **Stack-Level Tags** (sure, still without Tag definition within the template/on the resource). In my example below, I'm adding the following stack-level Tag:
```
TagFromCdkOnStack - 1
```
![alarm-stacktags-1](https://github.com/aws-cloudformation/cloudformation-coverage-roadmap/assets/16392819/ca92b088-c113-43bf-9f4e-1025300c848b)

2) Using the following simplified CFN YAML template as Stack-Template:

```yaml
---
AWSTemplateFormatVersion: '2010-09-09'
Description: CFN Drift detection test for AWS::CloudWatch::Alarm - test for new stack

Resources:
  SampleAlarmWithStackTags:
    Type: AWS::CloudWatch::Alarm
    Properties:
      AlarmDescription: We like Alarms with Tags and without Drifts
      AlarmName: TestAlarm-rogo
      ComparisonOperator: GreaterThanOrEqualToThreshold
      MetricName: HTTPCode_Target_5XX_Count
      Namespace: AWS/ApplicationELB
      EvaluationPeriods: 1
      Period: 900
      Statistic: Sum
      Threshold: 100
```

4) Checking the Console for Tags - We can see the stack-level Tag + built-in CFN tags / default service-managed auto-tags of CFN:
![alarm-stacktags-2](https://github.com/aws-cloudformation/cloudformation-coverage-roadmap/assets/16392819/3a9793bd-6778-4808-ab8f-f3353e72d4ee)

### Result
* If you create a _new_ stack and add stack-level Tags (it's a Best Practices following the AWS Cloud Development Kit (CDK) to add Tags on Stack-Level as well: https://docs.aws.amazon.com/cdk/v2/guide/tagging.html), CFN propagate the Stack-Tags to the `AWS::CloudWatch::Alarm` resource. (If you run Drift-Detection, Stack is IN_SYNC)
* If you have an _existing_ stack with stack-level Tags without an update, there are no tags. Furthermore those Stacks are in state DRIFTED.
* If an update of _existing_ stack with stack-level Tags would help, I need to check.


**Things are moving forward!** 🛼🛼 - I guess/hope it's an interim step before `AWS::CloudWatch::Alarm` will get `Tags` property on resource-level ❤️

**Comment by r-heimann** [comment]
I also just tested this in `eu-central-1`, and it seems like they sneakily !added! tag support. This works:

```yaml
AWSTemplateFormatVersion: 2010-09-09
Description: CFN Drift detection test for AWS::CloudWatch::Alarm - test for new stack

Resources:
  SampleAlarmWithStackTags:
    Type: AWS::CloudWatch::Alarm
    Properties:
      AlarmDescription: We like Alarms with Tags and without Drifts
      AlarmName: TestAlarm
      ComparisonOperator: GreaterThanOrEqualToThreshold
      MetricName: RunningInstancesCount
      Namespace: AWSLicenseManager/licenseUsage
      Dimensions:
       - Name: SubscriptionName
         Value: "Red Hat Enterprise Linux Server"
      EvaluationPeriods: 1
      Period: 900
      Statistic: Sum
      Threshold: 100
      Tags:
       - Key: Name
         Value: Test
```

![image](https://github.com/aws-cloudformation/cloudformation-coverage-roadmap/assets/121236014/4933180e-a503-41e6-a162-68b69d67e69d)

Thanks a lot to the CloudWatch devs, but for the love of god don't disable it 😂
The documentation is not updated yet: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cloudwatch-alarm.html
Now its time to _finally_ add tags to all of my Alarms.

**Comment by r-heimann** [comment]
> Hi all, I have to say you are an amazing community. Your votes helped us prioritize, and as you found out 😄 we are actively working on it. May I suggest to please hold a few more days until we can announce it before using it, as the deployment is still ongoing? In progress deployment implies no documentation as pointed out in the previous message, and a lot of regions still don't have it. But more importantly, if we hit any last minute issue with deployment we cannot fully rule out yet the risk of having to roll back and starting a new deployment again a few days later. I am really impatient to share this with you, with updated doc and a due announcement. I'll update here as soon as we are in a position to confirm the deployment is complete (expect a few more days). Thanks again for your great support and interest in this feature.

Thank you so much!

**Comment by farski** [comment]
After 5 years, literally yesterday I finally carved out some time to deployed a custom resource for tagging alarms across our AWS Organization. This is great news, but the timing 😭

**Comment by theaussiepom** [comment]
Oh man, I feel you farski!!! Fantastic it’s being worked on, but I too finally got a working solution up and running today, throttling retries and all 😂

**Comment by bdoyle0182** [comment]
So just to confirm what's coming in the next few days will only apply to stack level tags propagating to alarms correctly? Or individual alarm resource tags as well?

**Comment by iainelder** [comment]
@bdoyle0182 From @r-heimann's message it looks like alarm resource tags are on the way.

Let's be patient so @j-schwerer can complete the deployment. We've waited 5 years so 5 more days won't kill us :-)

Thanks to the CloudWatch and CloudFormation service teams for listening to the community!

**Comment by j-schwerer** [comment]
Hi all, yes indeed individual alarm resource tags are in scope as well. I'm as impatient as you to see this deployment completed. 
 :-) Stay tuned, so far so good

**Comment by bdoyle0182** [comment]
Amazing! So happy to see this being added

**Comment by r-heimann** [comment]
> Hi all, yes indeed individual alarm resource tags are in scope as well. I'm as impatient as you to see this deployment completed. :-) Stay tuned, so far so good

Just wanted to let you know that (in `eu-central-1`) the following works:

- Adding tags using `Tags` property
- Adding additional tags using `Tags` property
- Removing tags using `Tags` property
- Adding tags through Stack-Level Tags
- Adding tags through Stack-Level Tags + `Tags` property at the same time
- CloudFormation changesets for changing tags
- Drift detection works properly (original reason why we saw that something is being worked on, drift detection was broken for CloudWatch Alarms with an error i never saw before 😄)
- All CloudFormation default tags (`aws:cloudformation:*`) are properly added to the Alarms

This feels like an awesome, out-of-the-box working implementation. Although I'm only talking about "pure" CloudFormation, not entirely sure what has to be done for CDK support.

**Comment by j-schwerer** [comment]
Hi all, many many thanks for the great interaction here. It is my pleasure and privilege to confirm you that you can officially use it now, expect what's new post and doc updates in the coming hours

**Comment by tommyhegarty** [comment]
Re: @r-heimann 's comment; when should we expect CDK support to roll out?

**Comment by tmcclarnon** [comment]
We have also seen an issue where only tags that have new values are being applied to the alarm resources.

**Comment by kddejong** [comment]
As this is supported we are going to close this issue.  We can work on new issues as needed. https://aws.amazon.com/about-aws/whats-new/2024/04/amazon-cloudwatch-tagging-alarms-cloudformation/

**Comment by kddejong** [comment]
> Re: @r-heimann 's comment; when should we expect CDK support to roll out? Currently, using best-practices CDK (making use of `Tags.of(stack).add(key, value)`) does not propagate the tags down to the individual Alarm resources when `cdk synth`ed, at least when running version `2.135`

This should come shortly.  CDK works off of the specs/schemas so they will be building out the layer 1 tag support shortly as those files are updated with the tagging property.

**Comment by rgoltz** [comment]
> Re: @r-heimann 's comment; when should we expect CDK support to roll out? Currently, using best-practices CDK (making use of `Tags.of(stack).add(key, value)`) does not propagate the tags down to the individual Alarm resources when `cdk synth`ed, at least when running version `2.135`

If you are using higher-level constructs (e.g. L2), it needs implementation/a PR: https://github.com/aws/aws-cdk/issues/29759 (👍 would be good as well.) In case you are using L1 in CDK, it should get available soon after CDK L1 CloudFormation resource definitions update as part of an upcoming CDK release.

**Comment by katlim-br** [comment]
(1: And I thought Alarms supported tags from way before Apr 2024)
(2: I ask here because it seems like the only place where I can find some related information)

I am creating a CW Alarm with Tags using CloudFormation. It works! However, when the alarm is triggered and sent to the SNS topic, the Tags property does not arrive.

I just want to know if this is by design or could it be a missing piece or bug?

When I configure the Subscription, I also see a checkbox "Enable raw delivery" but that doesn't seem to make a difference.

Hopefully I can get an answer 🤞

**Comment by j-schwerer** [comment]
Hi Kat Lim, thanks for the question. Alarms don't propagate the tags to the notifications. Some customers use a Lambda to create a customized notification, including enriching the notification with a subset of the alarm tags, when they need. As this thread is closed, would it make sense to take future questions that don't fit in CloudFormation roadmap to the AWS repost site?
